introducing-pgzx.mdx

```diff
@@ -0,0 +1,208 @@
+
+Another advantage of memory contexts is that they improve memory monitoring, because context have names and you can see how much memory is used by each context. This is useful for debugging memory leaks, for example.
+
+A nice thing about the Zig standard library is that all functions/objects that allocate memory receive an allocator as a parameter. This makes allocations [more explicit](https://notes.eatonphil.com/2024-03-15-zig-rust-and-other-languages.html) but also makes it easy to use a custom allocator. pgzx defines custom allocators that wrap the Postgres memory contexts and make them available to Zig code.
+
+Here is an example that creates a new context as a child of the current context and gets the allocator for it:
+
+```
+var memctx = try pgzx.mem.createAllocSetContext("zig_context", .{ .parent = pg.CurrentMemoryContext });
+const allocator = memctx.allocator();
+```
+
+### Error handling
+
+Another Postgres API that you will very likely need to know about in a more complex extension is the error handling. Postgres implements "exceptions" in C via `setjmp/longjmp` and provides a set of macros to throw and catch them ([PG_TRY/PG_CATCH](https://github.com/postgres/postgres/blob/master/src/include/utils/elog.h#L318)).
+
+The issue is that long jumps could bypass the Zig control flow, for example `errdefer` blocks might not be executed. This means that if your extension calls into Postgres APIs, and those APIs can throw errors, long jumps might skip your `defer` and `errdefer` blocks!
+
+Luckily, pgzx is here to help. It provides a set of functions that allow you to catch the Postgres exceptions and convert them into Zig errors. For example:
+
+```
+var errctx = pgzx.err.Context.init();
+defer errctx.deinit();
+if (errctx.pg_try()) {
+    // zig code that calls several Postgres C functions.
+} else {
+    return errctx.errorValue();
+}
+```
+
+### Development Environment
+
+pgzx comes with a [Nix flakes](https://nixos.wiki/wiki/Flakes) based development environment for developing extensions as well as pgzx itself. It also comes with a project template which you can use to set up this environment in a new repository. To use it, [install Nix](https://github.com/DeterminateSystems/nix-installer) and then run:
+
+```bash
+mkdir my_extension
+cd my_extension
+nix flake init -t github:xataio/pgzx
+```
+
+Then load the nix shell with:
+
+```bash
+nix develop
+```
+
+The dev environment includes commands to relocate the Postgres source code, to start the server, and so on. The template also comes with a minimal extension and a `build.zig` file with a few common tasks. See the template [README](https://github.com/xataio/pgzx/tree/main/nix/templates/init) for how to build the extension form this point.
```

Comment:
We relocate the Nix Postgres binaries into your local development environment. This is much faster then building Postgres. We use that setup with relocating our binaries in our github workflow as well. The template does not create github workflow files, but one might still want to have a look at our workflow files if they want to Github Actions for CI.

We also have a script to fetch and create a Postgres debug build in you development environment. That later is great if you want to inspect Postgres source code or use a debugger and step through the Zig and Postgres code in your debugger. Because we use meson/Ninja to create the debug build you will also get a `compile_commands.json` file that should work with your editors C LSP ;)

We also have a similar script for the Zig toolchain. More details in our Readme.

Reply:
Corrected.